optuna/study/_multi_objective.py

```diff
@@ -127,102 +76,138 @@ def _fast_non_dominated_sort(
+        return on_front
+    elif n_objectives == 2:
+        return _is_pareto_front_2d(unique_lexsorted_loss_values)
+    else:
+        return _is_pareto_front_nd(unique_lexsorted_loss_values)
+
+
+def _is_pareto_front(loss_values: np.ndarray, assume_unique_lexsorted: bool = True) -> np.ndarray:
+    if assume_unique_lexsorted:
+        return _is_pareto_front_for_unique_sorted(loss_values)
+
+    unique_lexsorted_loss_values, order_inv = np.unique(loss_values, axis=0, return_inverse=True)
+    on_front = _is_pareto_front_for_unique_sorted(unique_lexsorted_loss_values)
+    return on_front[order_inv]
 
 
 def _calculate_nondomination_rank(
-    objective_values: np.ndarray,
-    *,
-    n_below: int | None = None,
-    base_rank: int = 0,
-) -> tuple[np.ndarray, int]:
-    if n_below is not None and n_below <= 0:
-        return np.full(len(objective_values), -1), base_rank
-    # Normalize n_below.
-    n_below = n_below or len(objective_values)
-    n_below = min(n_below, len(objective_values))
-
-    # The ndarray `domination_mat` is a boolean 2d matrix where
-    # `domination_mat[i, j] == True` means that the j-th trial dominates the i-th trial in the
-    # given multi objective minimization problem.
-    domination_mat = np.all(
-        objective_values[:, np.newaxis, :] >= objective_values[np.newaxis, :, :], axis=2
-    ) & np.any(objective_values[:, np.newaxis, :] > objective_values[np.newaxis, :, :], axis=2)
-
-    domination_list = np.nonzero(domination_mat)
-    domination_map = defaultdict(list)
-    for dominated_idx, dominating_idx in zip(*domination_list):
-        domination_map[dominating_idx].append(dominated_idx)
-
-    ranks = np.full(len(objective_values), -1)
-    dominated_count = np.sum(domination_mat, axis=1)
-
-    rank = base_rank - 1
-    ranked_idx_num = 0
-    while ranked_idx_num < n_below:
-        # Find the non-dominated trials and assign the rank.
-        (non_dominated_idxs,) = np.nonzero(dominated_count == 0)
-        ranked_idx_num += len(non_dominated_idxs)
-        rank += 1
-        ranks[non_dominated_idxs] = rank
+    loss_values: np.ndarray, *, n_below: int | None = None
+) -> np.ndarray:
+    if len(loss_values) == 0 or (n_below is not None and n_below <= 0):
+        return np.zeros(len(loss_values), dtype=int)
 
-        # Update the dominated count.
-        dominated_count[non_dominated_idxs] = -1
-        for non_dominated_idx in non_dominated_idxs:
-            dominated_count[domination_map[non_dominated_idx]] -= 1
+    (n_trials, n_objectives) = loss_values.shape
+    if n_objectives == 1:
+        _, ranks = np.unique(loss_values[:, 0], return_inverse=True)
+        return ranks
+    else:
+        # It ensures that trials[j] will not dominate trials[i] for i < j.
+        # np.unique does lexsort.
+        unique_lexsorted_loss_values, order_inv = np.unique(
+            loss_values, return_inverse=True, axis=0
+        )
```

Comment:
```suggestion
    # It ensures that trials[j] will not dominate trials[i] for i < j.
    # np.unique does lexsort.
    unique_lexsorted_loss_values, order_inv = np.unique(
        loss_values, return_inverse=True, axis=0
    )
```
You can omit this `else` block.

Reply:
You are right:)